dhis2/dhis2-core
# feat: add data sharing for SQL views [DHIS2-7410]

### Summary

Adds data sharing to SQL views.

This mean viewing the result grid of the SQL view requires data sharing access.
This has to be granted explicitly in the maintenance configuration.

The update of existing `Sharing` columns was extracted so it can be reused in case a similar case occurs again using:
```java
String updatedSharing = SharingUtils.withAccess( sharing, Sharing::copyMetadataToData );
```
The added `copyMetadataToData` method used here overrides any data sharing with the metadata sharing, in other words `ab??????` becomes `abab????` (all remaining `?` remain unchanged). Which means 
* `rw??????` => `rwrw????`
* `r-??????` => `r-r-????`
* ...

### Automatic Testing
A new spring test case was added to make sure viewing the grid requires special rights.
Existing tests were adopted to get past the rights check where needed.
Test setup was improved so that the `CurrentUserService` did no longer need mocking and updating in the tested service using a setter.

Tests were added to verify the behaviour of `Sharing` utility methods to set users and groups.

In addition there are several tests in connection with the changes made to perform the `Sharing` related DB updates.
The added `Sharing#withAccess` method is unit tested for correctness.

The DB JSON update it also unit tested to make sure the `String` => `Sharing` => `Sharing` => `String` transformation itself works. 

### Manual Testing

* start with a fresh Sierra Leone sample database
* login as admin
* open _Maintenance_ app
* select _SQL View_ in group _Other_ (http://localhost:8080/dhis-web-maintenance/index.html#/list/otherSection/sqlView)
* click the `...` button of one of the list entries (e.g. _Data elements_) and select _Sharing Settings_
* click the edit pen for the _Public Access_ entry
* with the change there now should be a section _DATA_, select _Can view only_
